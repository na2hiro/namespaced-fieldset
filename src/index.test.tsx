import React from "react";
import {describe, expect, it} from "vitest";
import {render} from "@testing-library/react";

import {constructFromFormData, constructFromQueryString, Fieldset, Input, namespaceAware} from ".";

describe("<Fieldset />", () => {
    it("renders fieldset", () => {
        const {container} = render(<Fieldset namespace="topLevel">
            <Input name="attribute" value="value"/>
            <Input name="others" value="foobar"/>
        </Fieldset>);
        expect(container).toMatchSnapshot();
    });
    it("adds namespaces with nesting", () => {
        const {container} = render(<Fieldset namespace="topLevel">
            <Input name="attribute" value="outer value"/>
            <Fieldset namespace={"secondLevel"}>
                <Input name="attribute" value="inner value"/>
            </Fieldset>
        </Fieldset>);
        expect(container).toMatchSnapshot();
    });
    it("accepts deeper namespaces with array and dot notations", () => {
        const {container} = render(<form>
            <Fieldset namespace="person[0]">
                <Input name="name" value="Jane"/>
                <Fieldset namespace={"hobby.pet"}>
                    <Input name="kind" value="koala"/>
                </Fieldset>
            </Fieldset>
            <Fieldset namespace="person[1]">
                <Input name="name" value="Jack"/>
                <Fieldset namespace={"hobby"}>
                    <Input name="favorite.cities[0]" value="Kyoto"/>
                </Fieldset>
            </Fieldset>
        </form>);
        expect(container).toMatchSnapshot();
    });
});

describe("<Fieldset.Headless />", () => {
    it("doesn't render fieldset", () => {
        const {container} = render(<Fieldset.Headless namespace="topLevel">
            <Input name="attribute" value="value"/>
            <Input name="others" value="foobar"/>
        </Fieldset.Headless>);
        expect(container).toMatchSnapshot();
    });
    it("adds namespaces with nesting, mixed with head-ful ones", () => {
        const {container} = render(<Fieldset.Headless namespace="topLevel">
            <Input name="foo" value="top-level value"/>
            <Fieldset.Headless namespace={"secondLevelHeadless"}>
                <Input name="bar" value="headless second level"/>
            </Fieldset.Headless>
            <Fieldset namespace={"secondLevelHeadful"}>
                <Input name="baz" value="headful second level"/>
            </Fieldset>
        </Fieldset.Headless>);
        expect(container).toMatchSnapshot();
    });
});

describe("<Input>", () => {
    it("works without wrapping with Fieldset", () => {
        const {container} = render(<form>
            <Input name="name" value="value"/>
            <Input type="submit" value="Proceed"/>
        </form>);
        expect(container).toMatchSnapshot();
    });
});

describe("namespaceAware()", () => {
    it("turns a component which takes name into a namespace aware one", () => {
        const Component = (props: { name: string }) => <div>gorgeous UI here! <input name={props.name}/></div>;
        const NamespacedComponent = namespaceAware(Component);
        const {container} = render(<Fieldset namespace="topLevel"><NamespacedComponent name="value"/></Fieldset>);
        expect(container).toMatchSnapshot();
    });
})

describe("constructFromFormData", () => {
    it("works with a simple query string", () => {
        const {container} = render(<form>
            <Fieldset namespace="person[0]">
                <Input name="name" value="Jane"/>
                <Fieldset namespace={"hobby.pets[0]"}>
                    <Input name="kind" value="koala"/>
                </Fieldset>
                <Fieldset namespace={"hobby.pets[1]"}>
                    <Input name="kind" value="cat"/>
                </Fieldset>
            </Fieldset>
            <Fieldset namespace="person[1]">
                <Input name="name" value="Jack"/>
                <Fieldset namespace={"hobby"}>
                    <Input name="favorite.cities" value="Kyoto"/>
                    <Input name="favorite.cities" value="Tokyo"/>
                </Fieldset>
            </Fieldset>
        </form>);
        const formElement = container.getElementsByTagName("form")[0] as HTMLFormElement;
        console.log(formElement.tagName)
        const formData = new FormData(formElement);
        expect(constructFromFormData(formData)).toMatchInlineSnapshot(`
          {
            "person": [
              {
                "hobby": {
                  "pets": [
                    {
                      "kind": "koala",
                    },
                    {
                      "kind": "cat",
                    },
                  ],
                },
                "name": "Jane",
              },
              {
                "hobby": {
                  "favorite": {
                    "cities": [
                      "Kyoto",
                      "Tokyo",
                    ],
                  },
                },
                "name": "Jack",
              },
            ],
          }
        `);
    });
})

describe("constructFromQueryString", () => {
    it("works with a simple query string", () => {
        expect(constructFromQueryString("person[0].name=bar&person[1].name=qux")).toMatchInlineSnapshot(`
          {
            "person": [
              {
                "name": "bar",
              },
              {
                "name": "qux",
              },
            ],
          }
        `);
    });
})