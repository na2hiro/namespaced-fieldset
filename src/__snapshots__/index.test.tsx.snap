// Vitest Snapshot v1

exports[`<Fieldset /> > accepts deeper namespaces with array and dot notations 1`] = `
<div>
  <form>
    <fieldset>
      <input
        name="person[0].name"
        value="Jane"
      />
      <fieldset>
        <input
          name="person[0].hobby.pet.kind"
          value="koala"
        />
      </fieldset>
    </fieldset>
    <fieldset>
      <input
        name="person[1].name"
        value="Jack"
      />
      <fieldset>
        <input
          name="person[1].hobby.favorite.cities[0]"
          value="Kyoto"
        />
      </fieldset>
    </fieldset>
  </form>
</div>
`;

exports[`<Fieldset /> > adds namespaces with nesting 1`] = `
<div>
  <fieldset>
    <input
      name="topLevel.attribute"
      value="outer value"
    />
    <fieldset>
      <input
        name="topLevel.secondLevel.attribute"
        value="inner value"
      />
    </fieldset>
  </fieldset>
</div>
`;

exports[`<Fieldset /> > renders fieldset 1`] = `
<div>
  <fieldset>
    <input
      name="topLevel.attribute"
      value="value"
    />
    <input
      name="topLevel.others"
      value="foobar"
    />
  </fieldset>
</div>
`;

exports[`<Fieldset.Headless /> > adds namespaces with nesting, mixed with head-ful ones 1`] = `
<div>
  <input
    name="topLevel.foo"
    value="top-level value"
  />
  <input
    name="topLevel.secondLevelHeadless.bar"
    value="headless second level"
  />
  <fieldset>
    <input
      name="topLevel.secondLevelHeadful.baz"
      value="headful second level"
    />
  </fieldset>
</div>
`;

exports[`<Fieldset.Headless /> > doesn't render fieldset 1`] = `
<div>
  <input
    name="topLevel.attribute"
    value="value"
  />
  <input
    name="topLevel.others"
    value="foobar"
  />
</div>
`;

exports[`<Input> > works without wrapping with Fieldset 1`] = `
<div>
  <form>
    <input
      name="name"
      value="value"
    />
    <input
      type="submit"
      value="Proceed"
    />
  </form>
</div>
`;

exports[`namespaceAware() > turns a component which takes name into a namespace aware one 1`] = `
<div>
  <fieldset>
    <div>
      gorgeous UI here! 
      <input
        name="topLevel.value"
      />
    </div>
  </fieldset>
</div>
`;
